matrix_sdk/src/room/joined.rs

```diff
@@ -0,0 +1,520 @@
+use crate::{room::Common, Client, Error, Result, Room, RoomType};
+use std::{io::Read, ops::Deref, sync::Arc};
+
+use matrix_sdk_common::{
+    api::r0::{
+        membership::{
+            ban_user,
+            invite_user::{self, InvitationRecipient},
+            join_room_by_id, join_room_by_id_or_alias, kick_user, Invite3pid,
+        },
+        message::send_message_event,
+        read_marker::set_read_marker,
+        receipt::create_receipt,
+        room::create_room,
+        typing::create_typing_event::{Request as TypingRequest, Typing},
+    },
+    assign,
+    events::{
+        room::{
+            message::{
+                AudioMessageEventContent, FileMessageEventContent, ImageMessageEventContent,
+                MessageEventContent, MessageType, VideoMessageEventContent,
+            },
+            EncryptedFile,
+        },
+        AnyMessageEventContent,
+    },
+    identifiers::{EventId, RoomId, RoomIdOrAliasId, ServerName, UserId},
+    uuid::Uuid,
+};
+
+use mime::{self, Mime};
+
+#[cfg(feature = "encryption")]
+use matrix_sdk_common::locks::Mutex;
+
+#[cfg(feature = "encryption")]
+use matrix_sdk_base::crypto::AttachmentEncryptor;
+
+#[cfg(feature = "encryption")]
+use tracing::instrument;
+
+/// A room in the joined state.
+///
+/// The `JoinedRoom` contains all methodes specific to a `Room` with type `RoomType::Joined`.
+/// Operations may fail once the underlaying `Room` changes `RoomType`.
+#[derive(Debug, Clone)]
+pub struct Joined {
+    inner: Common,
+}
+
+impl Deref for Joined {
+    type Target = Common;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl Joined {
+    /// Create a new `room::Joined` if the underlaying `Room` has type `RoomType::Joined`.
+    ///
+    /// # Arguments
+    /// * `client` - The client used to make requests.
+    ///
+    /// * `room` - The underlaying room.
+    pub fn new(client: Client, room: Room) -> Option<Self> {
+        if room.room_type() == RoomType::Joined {
+            Some(Self {
+                inner: Common::new(client, room),
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Join a room by `RoomId`.
+    ///
+    /// Returns a `join_room_by_id::Response` consisting of the
+    /// joined rooms `RoomId`.
+    ///
+    /// # Arguments
+    ///
+    /// * `room_id` - The `RoomId` of the room to be joined.
+    pub async fn join_by_id(client: Client, room_id: &RoomId) -> Result<join_room_by_id::Response> {
```

Comment:
Why are we trying to join joined rooms?

Reply:
this is a constructor for a joined room. Ultimately this should return a new `room::Joined`.

Reply:
Ah, in that case it shouldn't be `pub`.

Reply:
Why not? The user may want to join a new room by id.

Reply:
They may, but this should be a method under the `Client` object, so `client.join_room_by_id(room_id)` instead of `JoinedRoom::join_by_id(client, room_id)`.

Reply:
yes, if we make other constructors private then it makes much more sense to move it to the client.

Reply:
I think it makes sense either way, the pass by value style that's proposed here is less discoverable and slightly harder to use than this being a method under the client.

Reply:
i think it depends on the POV, but fine with me :)